### PR TITLE
`cedana manage`, runc GPU support

### PR DESCRIPTION
**Pull request by yashanand1910**
### Main changes
- New `cedana manage` subcommand has been added. It can be used to start managing a process, or runc container that wasn't started by the daemon. A managed process/runc container becomes a ‘job’, and allows C/Ring process and runc containers using just a job ID. Once a process or runc container becomes a managed job, its lifecycle is tied to the daemon. The daemon will keep track of the job, and will kill it during shut down.
- `cedana manage` supports the `--gpu-enabled` flag, thus enabling GPU C/R support for runc containers. It assumes the runc container is running with GPU LD_PRELOAD. In a later PR, we can create a runc hook/alternative that sets up GPU LD_PRELOAD for the container. This would be a separate binary/script.
- To support above work, `cedana ps` now emits more information, as shown below:
image
- `--gpu-enabled` flag for `cedana dump` has been removed. If a job was started with `--gpu-enabled`, the dump will automatically include a GPU checkpoint. Because, it doesn't make sense to pass this flag for a non-GPU job anyways. `ProcessState` now has a `GPU` field, which is set to `true` if the job was started with `--gpu-enabled`.

### Other changes, reducing complexity
- `--working-dir` for `cedana exec` is optional now and set to current working directory if not specified.
- `--dir` is now an optional flag for all dump cmds, as the daemon uses a `/tmp` directory by default.
- `--bundle` flag for runc restore is now optional. If not provided, it uses `ContainerBundle`, if found in saved `ProcessState` in dump.
- The service client in `cmd/*` is now created in a parent command's `PersistentPreRunE`, to reduce redundant lines of code.
- `--pid` flag has been removed from runc/containerd dump, as it is redundant. The pid is being fetched from container ID anyways.
- UID, GID, and Groups has been removed from restore args, as they can be fetched from process state in dump.

All changes are in sync with `cedana-api` PR (https://github.com/cedana/cedana-api/pull/34). Unused fields have been removed.

## Checklist before requesting a review
- [x] Have I performed a self-review?
- [ ] Have I added regression or unit tests (where it makes sense) for my changes?
- [x] Have I updated the README if changes have resulted in it being out of date?
- [x] Have I checked to ensure my PR only introduces the changes it's intended to? E.g. No extraneous formatting changes.
- [ ] Have I ensured that there are no performance regressions by checking benchmark results?
- [ ] Is this a breaking change? If yes, please describe areas affected and request reviews from developer stakeholders.

### CODE REVIEW

File: cmd/dump.go [hunk 1]
```diff
@@ -257,11 +262,8 @@ var dumpContainerdCmd = &cobra.Command{
 		}
 
 		root, _ := cmd.Flags().GetString(rootFlag)
-
 		dir, _ := cmd.Flags().GetString(dirFlag)
 		wdPath, _ := cmd.Flags().GetString(wdFlag)
-		pid, _ := cmd.Flags().GetInt(pidFlag)
```
[the remark this answers is not in the document]
Comment (yashanand1910) [reply cmd/dump.go:263]: Got it. In that case, leaving it removed. Helps to reduce complexity, as it’s only going to increase in the near future.

File: pkg/api/runc.go [hunk 2]
```diff
@@ -23,30 +23,122 @@ import (
...
 
-	return false, nil
+	// Wait for server shutdown to gracefully exit, since job is now managed
+	// Also wait for process exit, to update state
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		select {
+		case <-s.serverCtx.Done():
+			<-s.serverCtx.Done()
```
[the remark this answers is not in the document]
Comment (yashanand1910) [reply pkg/api/runc.go:92]: Nice catch.

File: pkg/api/runc.go [hunk 3]
```diff
@@ -57,24 +149,45 @@ func (s *service) RuncDump(ctx context.Context, args *task.RuncDumpArgs) (*task.
...
+	isManagedJob := false // if a JID = ContainerID exists in DB
+	state, err := s.getState(ctx, args.ContainerID)
+	if err == nil {
+		isManagedJob = true
+		if state.GPU && s.gpuEnabled == false {
+			return nil, status.Error(codes.FailedPrecondition, "GPU support is not enabled in daemon")
+		}
+	}
+
+	state, err = s.generateState(ctx, pid)
```
[the remark this answers is not in the document]
Comment (yashanand1910) [reply pkg/api/runc.go:171]: Yeah, that’s right. The state is then only _saved_ if the job is managed.

File: .gitmodules [hunk 4]
```diff
@@ -1,4 +1,4 @@
 [submodule "pkg/api/services/cedana-api"]
 	path = pkg/api/services/cedana-api
 	url = https://github.com/cedana/cedana-api
-  branch = main
+  branch = feature/ced-455-cr-support-for-nvidia-container-runtime
```
[the remark this answers is not in the document]
Comment (yashanand1910) [reply .gitmodules:4]: Thanks for the reminder!

**Review by nravic — APPROVED** [review]
This looks good. While you're here, can you add a howto file to the docs folder (similar to the kata one) that walks through a simple usage of this? from starting a runc container -> cedana manage -> checkpoint -> restore.

**Pull request merged**